File: clojure/src/net_perspective/peer/batch.clj
(ns net-perspective.peer.batch
  "Inductive batch computation of context-relations-deps for all homed users."
  (:require [malli.core :as m]
            [net-perspective.schema :as schema]
            [net-perspective.ipfs.client :as ipfs]
            [net-perspective.peer.state :as state]
            [net-perspective.peer.registry :as registry])
  (:import [java.io ByteArrayOutputStream]
           [java.util.zip ZipOutputStream ZipEntry]))

;; ---------------------------------------------------------------------------
;; Document IO — validate on store and fetch

(defn- store-document!
  "Validates document against schema-var, marshals to JCS, stores in IPFS. Returns CID."
  [server document schema-var]
  (schema/validate! @schema-var document)
  (ipfs/add (:ipfs server) (schema/marshal-document document)))

(defn- fetch-document!
  "Fetches CID from IPFS, unmarshals, validates against schema-var."
  [server cid schema-var]
  (schema/validate! @schema-var (schema/unmarshal-document (ipfs/cat (:ipfs server) cid))))

;; ---------------------------------------------------------------------------
;; CID resolution (state lookups, no IPFS IO)

(defn- user-index-cid
  "Returns the index CID for a user-id (bytes or base64), or nil.
   Checks local state first, then the registry."
  [server user-id]
  (let [uid (schema/ensure-bytes user-id)]
    (or (some-> (state/get-user server uid) :index-cid not-empty)
        (some-> (:registry server) (registry/lookup uid) not-empty))))

(defn- find-deps-cid
  "Returns the crd-address for target-path in a parsed index, or nil."
  [index target-path]
  (some #(when (= (get % "crd-idx-ctx/path") target-path)
           (get % "crd-idx-ctx/crd-address"))
        (get index "crd-idx/contexts" [])))

;; ---------------------------------------------------------------------------
;; Pure document builders (adjust-hops, make-crd, find-deps-cid are tested directly)

(defn- adjust-hops
  "Increments hop counts by current-hop, dropping any that would exceed 10."
  [current-hop hops]
  (into []
        (comp (map #(update % "crd-hop/hop" + current-hop))
              (filter #(<= (get % "crd-hop/hop") 10)))
        hops))

(defn- make-crd
  "Builds a ContextRelationsDeps document."
  [user-id now context-path hops src-cids]
  {"crd/version"          1
   "crd/timestamp-ns"     now
   "crd/user-id"          user-id
   "crd/context-path"     context-path
   "crd/hops"             hops
   "crd/source-addresses" src-cids})

(defn- zip-archive ^bytes [entries]
  (let [baos (ByteArrayOutputStream.)]
    (with-open [zos (ZipOutputStream. baos)]
      (doseq [[name ^bytes data] entries]
        (.putNextEntry zos (ZipEntry. ^String name))
        (.write zos data)
        (.closeEntry zos)))
    (.toByteArray baos)))

;; ---------------------------------------------------------------------------
;; Fetch — read all data needed for one user's batch computation

(defn- fetch-related-deps
  "For each user-type relation in matching contexts, resolves and fetches their
   deps document. Returns a seq of {:deps doc, :deps-cid cid} maps."
  [server dr context-path current-hop]
  (let [relations (->> (get dr "dr/contexts" [])
                       (filter #(= (get % "dr-ctx/path") context-path))
                       (mapcat #(get % "dr-ctx/relations" []))
                       (filter #(= (get % "dr-rel/type") "user")))]
    (reduce
     (fn [acc rel]
       (let [max-depth (min 10 (max 2 (get rel "dr-rel-user/transitive-depth" 2)))]
         (if (>= current-hop max-depth)
           acc
           (let [rel-uid     (get rel "dr-rel-user/user-id")
                 target-path (or (seq (get rel "dr-rel-user/context-path")) context-path)]
             (if-let [idx-cid (user-index-cid server rel-uid)]
               (try
                 (let [index    (fetch-document! server idx-cid #'schema/ContextRelationsDepsIndex)
                       deps-cid (find-deps-cid index target-path)]
                   (if deps-cid
                     (conj acc {:deps     (fetch-document! server deps-cid #'schema/ContextRelationsDeps)
                                :deps-cid deps-cid})
                     acc))
                 (catch Exception _ acc))
               acc)))))
     []
     relations)))

(defn- fetch-user-data
  "Reads all data from IPFS needed for a user's batch computation.
   Returns a map with :dr-bytes, :arch-cid, and per-context :related-deps,
   or nil if the user has no DR to process."
  [server user]
  (let [dr     (:latest-dr user)
        dr-cid (:latest-dr-cid user)]
    (when (and dr (seq dr-cid))
      (let [dr-bytes (ipfs/cat (:ipfs server) dr-cid)
            arch-cid (ipfs/add (:ipfs server) (zip-archive [[dr-cid dr-bytes]]))]
        {:dr-bytes     dr-bytes
         :arch-cid     arch-cid
         :context-deps (into {}
                             (map (fn [ctx]
                                    [(get ctx "dr-ctx/path")
                                     (fetch-related-deps server dr (get ctx "dr-ctx/path") 1)]))
                             (get dr "dr/contexts" []))}))))

;; ---------------------------------------------------------------------------
;; Build — pure document creation from fetched data

(defn- build-context-crd
  "Builds a CRD document for one context from fetched data. Pure."
  [uid now dr-cid arch-cid dr-size related-deps context-path]
  (let [hop1       {"crd-hop/hop"             1
                    "crd-hop/dr-addresses"    [dr-cid]
                    "crd-hop/archive-address" arch-cid
                    "crd-hop/size"            dr-size}
        extra-hops (mapcat #(adjust-hops 1 (get (:deps %) "crd/hops" [])) related-deps)
        src-cids   (mapv :deps-cid related-deps)]
    (make-crd uid now context-path (into [hop1] extra-hops) src-cids)))

(defn- build-index-ctx-entry
  "Builds one index context entry from a CRD document and its CID. Pure."
  [cpath crd crd-cid]
  (let [dep-hops (get crd "crd/hops" [])]
    {"crd-idx-ctx/path"        cpath
     "crd-idx-ctx/crd-address" crd-cid
     "crd-idx-ctx/hops"        (reduce max 0 (map #(get % "crd-hop/hop" 0) dep-hops))
     "crd-idx-ctx/size"        (reduce + 0 (map #(get % "crd-hop/size" 0) dep-hops))}))

(defn- build-documents
  "Builds all documents for one user from fetched data. Pure.
   Returns {:crds [{:cpath :crd :idx-entry}], :index doc, :envelope doc}."
  [user fetched-data now]
  (let [dr      (:latest-dr user)
        dr-cid  (:latest-dr-cid user)
        kp      (:key-pair user)
        uid     (:user-id kp)
        dr-size (alength ^bytes (:dr-bytes fetched-data))

        crds (mapv (fn [ctx]
                     (let [cpath (get ctx "dr-ctx/path")
                           related (get (:context-deps fetched-data) cpath [])
                           crd   (build-context-crd uid now dr-cid (:arch-cid fetched-data)
                                                    dr-size related cpath)]
                       {:cpath cpath :crd crd}))
                   (get dr "dr/contexts" []))

        ;; index and envelope need CRD CIDs, which aren't known until store.
        ;; Return the documents; store-user-documents! will wire up the CIDs.
        user-info {"user/version"         1
                   "user/timestamp-ns"    now
                   "user/user-id"         uid
                   "user/user-public-key" (:encoded-public-key kp)
                   "user/dr-address"      dr-cid}]
    {:crds      crds
     :uid       uid
     :now       now
     :user-info user-info
     :kp        kp}))

;; ---------------------------------------------------------------------------
;; Store — persist built documents to IPFS and update state

(defn- store-user-documents!
  "Stores all built documents for one user. Returns {:index-cid :ui-cid}."
  [server user {:keys [crds uid now user-info kp]}]
  (let [idx-ctxs (mapv (fn [{:keys [cpath crd]}]
                          (let [crd-cid (store-document! server crd #'schema/ContextRelationsDeps)]
                            (build-index-ctx-entry cpath crd crd-cid)))
                        crds)
        index     {"crd-idx/version"      1
                   "crd-idx/timestamp-ns" now
                   "crd-idx/user-id"      uid
                   "crd-idx/contexts"     idx-ctxs}
        index-cid (store-document! server index #'schema/ContextRelationsDepsIndex)
        ui-cid    (store-document! server (schema/wrap-envelope user-info kp) #'schema/Envelope)]
    (state/set-index-cid! server uid index-cid)
    (ipfs/publish-ipns (:ipfs server) (:ipns-key-name user) ui-cid)))

;; ---------------------------------------------------------------------------
;; Orchestration

(defn- process-user! [server user]
  (when-let [fetched (fetch-user-data server user)]
    (let [docs (build-documents user fetched (System/nanoTime))]
      (store-user-documents! server user docs))))

;; ---------------------------------------------------------------------------
;; Public entry point

(defn run-batch!
  "Runs one batch update round for all homed users."
  [server]
  (doseq [user (state/all-users server)]
    (try (process-user! server user)
         (catch Exception e
           (println (str "batch: user error: " (.getMessage e)))))))
(m/=> run-batch! [:=> [:cat #'schema/PeerServer] :nil])

File: pkg/peer/batch.go
package peer

import (
	"context"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"time"

	"github.com/decentstates/net-perspective/pkg/doc"
)

// RunBatch computes context-relations-deps for all homed users, stores them to
// IPFS, and atomically publishes each user's updated user-info via IPNS.
//
// The final IPNS publish of each user-info is the atomic commit: a cancelled
// context leaves the previous consistent state intact.
func (s *Server) RunBatch(ctx context.Context) error {
	users := s.AllUsers()
	if len(users) == 0 {
		return nil
	}

	for _, u := range users {
		if err := ctx.Err(); err != nil {
			return err
		}
		if err := s.processUser(ctx, u); err != nil {
			// Log and continue; one user's failure shouldn't block others.
			log.Printf("batch: user %x: %v", u.KeyPair.UserID[:8], err)
		}
	}
	return nil
}

func (s *Server) processUser(ctx context.Context, u *HomedUser) error {
	dr, _ := u.DirectRelations()
	if dr == nil {
		return nil // nothing submitted yet
	}

	now := time.Now().UnixNano()

	// Build context-relations-deps per context and collect into an index.
	index := doc.ContextRelationsDepsIndex{
		Version:     1,
		TimestampNs: now,
		UserID:      u.KeyPair.UserID,
	}

	for _, ctx_ := range dr.Contexts {
		if err := ctx.Err(); err != nil {
			return err
		}

		deps, err := s.computeDeps(ctx, u, dr, ctx_.ContextPath)
		if err != nil {
			return fmt.Errorf("compute deps for %v: %w", ctx_.ContextPath, err)
		}

		depsBytes, err := doc.Marshal(deps)
		if err != nil {
			return err
		}
		depsCID, err := s.IPFS.Add(depsBytes)
		if err != nil {
			return fmt.Errorf("store deps: %w", err)
		}

		totalSize := int64(0)
		for _, h := range deps.Hops {
			totalSize += h.Size
		}
		maxHops := 0
		if len(deps.Hops) > 0 {
			maxHops = deps.Hops[len(deps.Hops)-1].Hop
		}

		index.Contexts = append(index.Contexts, doc.ContextRelationsDepsIndexContext{
			ContextPath:            ctx_.ContextPath,
			ContextRelsDepsAddress: []byte(depsCID),
			Hops:                   maxHops,
			Size:                   totalSize,
		})
	}

	// Store index.
	indexBytes, err := doc.Marshal(index)
	if err != nil {
		return err
	}
	indexCID, err := s.IPFS.Add(indexBytes)
	if err != nil {
		return fmt.Errorf("store index: %w", err)
	}
	u.SetIndexCID(indexCID)

	// Atomic commit: update user-info with the new index address and publish to IPNS.
	ui := &doc.UserInfo{
		Version:                         1,
		TimestampNs:                     now,
		UserID:                          u.KeyPair.UserID,
		UserPublicKey:                   u.KeyPair.EncodedPublicKey,
		ContextRelsDepsIndexIPNSAddress: indexCID,
	}
	// Preserve the direct-relations IPNS address if already set.
	if prev := u.latestUserInfo; prev != nil {
		ui.DirectRelationsIPNSAddress = prev.DirectRelationsIPNSAddress
		ui.TrustedPeers = prev.TrustedPeers
		ui.PeeredUsers = prev.PeeredUsers
	}

	env, err := doc.Wrap(ui, u.KeyPair)
	if err != nil {
		return fmt.Errorf("wrap user-info: %w", err)
	}
	uiBytes, err := doc.Marshal(env)
	if err != nil {
		return err
	}
	uiCID, err := s.IPFS.Add(uiBytes)
	if err != nil {
		return fmt.Errorf("store user-info: %w", err)
	}
	return s.IPFS.PublishIPNS(u.IPNSKeyName, uiCID)
}

// computeDeps builds a ContextRelationsDeps for one user+context by scanning
// the direct-relations transitively up to the per-relation transitive-depth cap.
//
// Fetching remote peers' deps and seeding is Phase 3+ work; for now it builds
// hop-1 from the user's own direct-relations.
func (s *Server) computeDeps(
	ctx context.Context,
	u *HomedUser,
	dr *doc.DirectRelations,
	contextPath []string,
) (*doc.ContextRelationsDeps, error) {
	_, drCID := u.DirectRelations()

	hop := doc.ContextRelationsDepsHop{
		Hop:                          1,
		DirectRelationsAddresses:     [][]byte{[]byte(drCID)},
		DirectRelationsArchiveAddress: []byte(drCID), // archive = same doc for hop-1
	}

	// Compute hop size from the stored direct-relations bytes.
	if drCID != "" {
		data, err := s.IPFS.Cat(drCID)
		if err == nil {
			hop.Size = int64(len(data))
		}
	}

	// Fetch and incorporate deps from related users (hop 2+).
	additionalHops, sourceAddrs, err := s.fetchTransitiveDeps(ctx, dr, contextPath, 1)
	if err != nil {
		// Non-fatal: log and continue with what we have.
		log.Printf("fetchTransitiveDeps: %v", err)
	}

	hops := append([]doc.ContextRelationsDepsHop{hop}, additionalHops...)

	return &doc.ContextRelationsDeps{
		Version:         1,
		TimestampNs:     time.Now().UnixNano(),
		UserID:          u.KeyPair.UserID,
		ContextPath:     contextPath,
		Hops:            hops,
		SourceAddresses: sourceAddrs,
	}, nil
}

// fetchTransitiveDeps fetches context-relations-deps from directly related users
// and returns additional hops and source CIDs.
func (s *Server) fetchTransitiveDeps(
	ctx context.Context,
	dr *doc.DirectRelations,
	contextPath []string,
	currentHop int,
) ([]doc.ContextRelationsDepsHop, [][]byte, error) {
	var hops []doc.ContextRelationsDepsHop
	var sourceAddrs [][]byte

	for _, c := range dr.Contexts {
		if !pathMatches(c.ContextPath, contextPath) {
			continue
		}
		for _, rel := range c.Relations {
			if rel.Type != "user" {
				continue
			}
			maxDepth := rel.TransitiveDepth
			if maxDepth == 0 {
				maxDepth = 2
			}
			if maxDepth > 10 {
				maxDepth = 10
			}
			if currentHop >= maxDepth {
				continue
			}

			// Resolve the related user's index.
			deps, depsCID, err := s.fetchUserContextDeps(ctx, rel, contextPath)
			if err != nil {
				log.Printf("fetchUserContextDeps %x: %v", rel.RelUserID[:min(8, len(rel.RelUserID))], err)
				continue
			}
			if depsCID != "" {
				sourceAddrs = append(sourceAddrs, []byte(depsCID))
			}

			// Merge their hops, incrementing hop numbers.
			for _, h := range deps.Hops {
				adjusted := h
				adjusted.Hop = currentHop + h.Hop
				if adjusted.Hop > 10 {
					break
				}
				hops = append(hops, adjusted)
			}
		}
	}
	return hops, sourceAddrs, nil
}

// fetchUserContextDeps resolves a related user's context-relations-deps-index
// and extracts the entry for the given context path.
//
// It first checks if the user is homed locally. If not, and a PeerRegistry is
// configured, it queries the remote peer over HTTP.
func (s *Server) fetchUserContextDeps(
	ctx context.Context,
	rel doc.DirectRelation,
	contextPath []string,
) (*doc.ContextRelationsDeps, string, error) {
	targetPath := contextPath
	if len(rel.RelContextPath) > 0 {
		targetPath = rel.RelContextPath
	}

	// Try local first.
	if relUser := s.GetUser(rel.RelUserID); relUser != nil {
		return s.fetchLocalUserContextDeps(relUser, targetPath)
	}

	// Try remote via peer registry.
	if s.Registry != nil {
		if peerURL, ok := s.Registry.Lookup(rel.RelUserID); ok {
			return s.fetchRemoteUserContextDeps(ctx, peerURL, rel.RelUserID, targetPath)
		}
	}

	return &doc.ContextRelationsDeps{}, "", nil
}

func (s *Server) fetchLocalUserContextDeps(
	relUser *HomedUser,
	targetPath []string,
) (*doc.ContextRelationsDeps, string, error) {
	indexCID := relUser.IndexCID()
	if indexCID == "" {
		return &doc.ContextRelationsDeps{}, "", nil
	}
	indexBytes, err := s.IPFS.Cat(indexCID)
	if err != nil {
		return nil, "", err
	}
	return extractDepsFromIndex(indexBytes, targetPath, s.IPFS.Cat)
}

// fetchRemoteUserContextDeps fetches a user's index from a remote peer via HTTP,
// then fetches the specific deps document for the target context path.
func (s *Server) fetchRemoteUserContextDeps(
	ctx context.Context,
	peerBaseURL string,
	userID []byte,
	targetPath []string,
) (*doc.ContextRelationsDeps, string, error) {
	// GET /status/users/{hexUserID} to find the index CID.
	statusURL := peerBaseURL + "/status/users/" + hex.EncodeToString(userID)
	req, _ := http.NewRequestWithContext(ctx, http.MethodGet, statusURL, nil)
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, "", fmt.Errorf("fetch remote status: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusNotFound {
		return &doc.ContextRelationsDeps{}, "", nil
	}
	if resp.StatusCode != http.StatusOK {
		return nil, "", fmt.Errorf("remote status: %s", resp.Status)
	}

	var st UserStatus
	if err := json.NewDecoder(resp.Body).Decode(&st); err != nil {
		return nil, "", fmt.Errorf("decode remote status: %w", err)
	}
	if st.IndexCID == "" {
		return &doc.ContextRelationsDeps{}, "", nil
	}

	// Fetch the index document via /cid/.
	catRemote := func(cid string) ([]byte, error) {
		url := peerBaseURL + "/cid/" + cid
		req, _ := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
		resp, err := http.DefaultClient.Do(req)
		if err != nil {
			return nil, err
		}
		defer resp.Body.Close()
		if resp.StatusCode != http.StatusOK {
			return nil, fmt.Errorf("GET %s: %s", url, resp.Status)
		}
		return io.ReadAll(resp.Body)
	}

	indexBytes, err := catRemote(st.IndexCID)
	if err != nil {
		return nil, "", fmt.Errorf("fetch remote index: %w", err)
	}
	return extractDepsFromIndex(indexBytes, targetPath, catRemote)
}

// extractDepsFromIndex decodes an index document, finds the entry matching
// targetPath, and fetches its deps document using the provided fetch function.
func extractDepsFromIndex(
	indexBytes []byte,
	targetPath []string,
	fetch func(cid string) ([]byte, error),
) (*doc.ContextRelationsDeps, string, error) {
	var index doc.ContextRelationsDepsIndex
	if err := doc.Unmarshal(indexBytes, &index); err != nil {
		return nil, "", err
	}
	for _, entry := range index.Contexts {
		if pathMatches(entry.ContextPath, targetPath) {
			depsCID := string(entry.ContextRelsDepsAddress)
			depsBytes, err := fetch(depsCID)
			if err != nil {
				return nil, depsCID, err
			}
			var deps doc.ContextRelationsDeps
			if err := doc.Unmarshal(depsBytes, &deps); err != nil {
				return nil, depsCID, err
			}
			return &deps, depsCID, nil
		}
	}
	return &doc.ContextRelationsDeps{}, "", nil
}

// pathMatches returns true if a equals b (exact match for now; glob handled by client).
func pathMatches(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

